File: backend/app/schemas/camera.py
"""
Camera schemas for API request/response validation.
"""
from datetime import datetime
from typing import Optional, Literal
from pydantic import BaseModel, Field


class CameraBase(BaseModel):
    """Base camera schema."""
    name: str = Field(..., min_length=1, max_length=100, description="Camera name")
    source: str = Field(..., min_length=1, max_length=500, description="File path or RTSP URL")
    source_type: Literal["file", "rtsp"] = Field(default="file", description="Source type")


class CameraCreate(CameraBase):
    """Schema for creating a camera."""
    pass


class CameraUpdate(BaseModel):
    """Schema for updating a camera."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    source: Optional[str] = Field(None, min_length=1, max_length=500)
    source_type: Optional[Literal["file", "rtsp"]] = None
    is_active: Optional[bool] = None


class CameraResponse(CameraBase):
    """Schema for camera response."""
    id: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/lib/models/camera.dart
/// Camera model
class Camera {
  final int id;
  final String name;
  final String source;
  final String sourceType;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  Camera({
    required this.id,
    required this.name,
    required this.source,
    required this.sourceType,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Camera.fromJson(Map<String, dynamic> json) {
    return Camera(
      id: json['id'],
      name: json['name'],
      source: json['source'],
      sourceType: json['source_type'],
      isActive: json['is_active'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'source': source,
      'source_type': sourceType,
      'is_active': isActive,
    };
  }
}

/// Camera creation request
class CameraCreate {
  final String name;
  final String source;
  final String sourceType;

  CameraCreate({
    required this.name,
    required this.source,
    this.sourceType = 'file',
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'source': source,
      'source_type': sourceType,
    };
  }
}
